### PR TITLE
Baseline reports (Baseline 3 of 3)

### PR DESCRIPTION
**Pull request by jeremyandrews**
# Baseline reports (3 of 3)

Builds on PR #660 (baseline processing). Adds the `--baseline-file` CLI option and integrates baseline comparison into all report outputs (HTML, JSON, Markdown).

When a baseline file is provided, each numeric metric is displayed alongside its delta from the baseline run (e.g., `1000 (+200)` or `50 (-10)`). Metrics are correlated by key (method + name for requests, transaction + name for transactions, etc.), so new or removed endpoints are handled gracefully — unmatched metrics are displayed without deltas.

### CLI option (`src/config.rs`)

Adds `--baseline-file <PATH>` option backed by `GooseDefault::BaselineFile`. Accepts a path to a JSON report from a previous Goose run.

### Metric structs wrapped with `Value<T>` (`src/report.rs`)

All numeric fields on report metric structs are changed from plain types to `Value<T>`:

- `RequestMetric`: `usize` → `Value<usize>`, `f32` → `Value<NullableFloat>` (7 fields)
- `ResponseMetric`: `usize` → `Value<usize>` (8 percentile fields)
- `CORequestMetric`: `f32` → `Value<NullableFloat>`, `usize` → `Value<usize>` (3 fields)
- `TransactionMetric`: similar wrapping (7 fields)
- `ScenarioMetric`: similar wrapping (7 fields)
- `ErrorMetric` (new struct): `GooseMethod` + `Value<usize>` occurrences

Each struct implements `DeltaTo` for field-by-field baseline application. `StatusCodeMetric` is left as plain strings (no meaningful numeric delta).

### `DeltaTo` trait (`src/metrics/delta.rs`)

New trait for metric structs to accept baseline deltas:

```rust
pub trait DeltaTo {
    fn delta_to(&mut self, other: &Self);
}
```

### Baseline loading and correlation (`src/metrics/common.rs`)

- `load_baseline_file()`: Opens a JSON report, deserializes through `Baseline*` intermediate structs (for forward compatibility), validates structural consistency, returns `ReportData<'static>`.
- `correlate_deltas()`: Generic function that matches current/baseline metric slices by a key function and calls `DeltaTo::delta_to` on matched pairs.
- `apply_baseline_deltas()`: Single function that applies correlation across all metric types (requests, responses, CO requests, CO responses, transactions, scenarios, errors).
- `Baseline*` deserialization structs: `BaselineRequestMetric`, `BaselineResponseMetric`, `BaselineCORequestMetric`, `BaselineTransactionMetric`, `BaselineScenarioMetric`, `BaselineErrorMetric`, `BaselineGooseMetrics`, `BaselineReportData` — each with a `From` impl that wraps plain values in `Value::Plain(...)`.
- Validation: request/response count consistency check, non-empty check.

### Report integration (`src/metrics.rs`, `src/report.rs`, `src/report/markdown.rs`)

- `GooseMetrics` now derives `Serialize` + `Deserialize` (replacing a manual `Serialize` impl) with `#[serde(skip)]` on `history`.
- `prepare_report_data()`: New method on `GooseAttack` that loads the baseline file if configured and delegates to `prepare_data_with_baseline`.
- HTML, JSON, and Markdown report writers all go through `prepare_report_data()`, gaining baseline support automatically.
- `format_value()`: Formats `Value<T>` as `"100"` or `"100 (+20)"` for HTML template interpolation.
- Response percentiles use `format_value()` in both HTML and Markdown templates.

### Supporting changes

- `GooseMethod` gains `Copy` and `Hash` derives (`src/goose.rs`) — required for `ErrorMetric` and `correlate_deltas` key hashing.
- `TestPlanStepAction` and `TestPlanHistory` gain `Serialize`/`Deserialize` (`src/test_plan.rs`) — required for `GooseMetrics` derive.
- `chrono` gains the `"serde"` feature (`Cargo.toml`) — required for `GooseMetrics` derive.

## Files changed (9 files, +1142/-422)

| File | Lines | What |
|------|-------|------|
| `src/metrics/common.rs` | +894/-205 | Baseline loading, correlation, `Baseline*` structs, validation, tests |
| `src/report.rs` | +155/-61 | `Value<T>` wrapping, `DeltaTo` impls, `ErrorMetric`, `format_value` usage |
| `src/metrics.rs` | +68/-57 | `prepare_report_data()`, `format_value()`, derive Serialize on GooseMetrics |
| `src/metrics/delta.rs` | +10/-58 | `DeltaTo` trait, remove unused `formatted_number` |
| `src/report/markdown.rs` | +14/-11 | `format_value` for percentiles, `ErrorMetric` for error table |
| `src/config.rs` | +23/0 | `--baseline-file` CLI option |
| `src/test_plan.rs` | +2/-2 | Add Serialize/Deserialize derives |
| `src/goose.rs` | +1/-1 | Add Copy, Hash to GooseMethod |
| `Cargo.toml` | +1/-1 | chrono serde feature |

## Tests

5 unit tests covering the new baseline functionality:

- `delta_value_usize` — overflow-safe usize delta with edge cases (0, MAX)
- `correlate_deltas_matches_by_key` — verifies key-based matching; unmatched metrics stay `Plain`
- `load_baseline_file_invalid_path` — nonexistent file returns error
- `load_baseline_file_invalid_json` — malformed JSON returns error
- `load_baseline_file_empty_metrics` — valid JSON with zero metrics returns `InvalidOption` with descriptive message

## Design decisions

**Why `Baseline*` intermediate structs instead of deserializing directly into `Value<T>`?** Forward compatibility. `Value<T>` serializes as `{"value": 100, "delta": 50}` for deltas, but baseline files from older Goose versions will have plain `100`. The intermediate structs accept plain values and convert via `From` impls. If the JSON format changes in the future, only the `Baseline*` structs need updating.

**Why `Cow<'m, GooseMetrics>` in `ReportData`?** Live test data borrows `&GooseMetrics` from the running `GooseAttack`. Baseline data must own its `GooseMetrics` (loaded from file). `Cow` handles both without cloning the live case.

**Why a single `apply_baseline_deltas` instead of inline in each builder?** Avoids duplicating correlation logic across `build_raw`, `build_co`, `build_transaction`, `build_scenario`, and `build_errors`. The `Prepare` struct only builds current-run metrics; baseline deltas are applied once afterward.

**Why `GooseMethod` on `ErrorMetric` instead of `String`?** Baseline files serialize `GooseMethod` with serde's default enum representation. Using `GooseMethod` directly leverages derive `Deserialize` instead of manual string matching, and enables `Copy` + `Hash` for correlation keys.

## Review feedback addressed

| Feedback | Resolution |
|----------|-----------|
| Duplicated code between `Prepare`/`PrepareOwned` | Eliminated. Single `Prepare` struct, single `apply_baseline_deltas()` function. `prepare_data_with_baseline_owned` is a one-line wrapper. |
| Weak tests that only check config creation | Replaced with unit tests that exercise `load_baseline_file` and `correlate_deltas` directly, validating actual parsing and error handling. |
| Missing CO metric deltas | Added `co_request_metrics` and `co_response_metrics` correlation in `apply_baseline_deltas()`. |

### CODE REVIEW

File: src/metrics/common.rs [hunk 1]
```diff
@@ -38,111 +45,747 @@ pub struct ReportOptions {
...
+        let mut aggregate_scenario_times: BTreeMap<usize, usize> = BTreeMap::new();
+        let mut aggregate_iterations = 0.0;
+        let mut aggregate_response_time_counter = 0.0;
+        for scenario in &self.metrics.scenarios {
+            let (count_per_second, _failures_per_second) =
+                per_second_calculations(self.metrics.duration, scenario.counter, 0);
+            let average = match scenario.counter {
+                0 => 0.00,
+                _ => scenario.total_time as f32 / scenario.counter as f32,
+            };
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply src/metrics/common.rs]: good advice, the code has been deduplicated

File: src/metrics/delta.rs [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+                // ... which fits, so we return the negative value
+                -(delta as isize)
+            }
+        }
+    }
+
+    fn is_delta_positive(value: Self::Delta) -> bool {
+        value.is_positive()
+    }
+}
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply src/metrics/delta.rs:44]: "Thanks for the suggestion! The Hash trait has been implemented for Value with proper discriminant handling in lines 77-89 of src/metrics/delta.rs."

File: tests/baseline.rs [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+            "1",
+            "--run-time",
+            "1",
+            "--baseline-file",
+            temp_path,
+        ],
+    );
+
+    // Verify the baseline file was set through CLI parsing
+    assert!(config.baseline_file.is_some());
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply tests/baseline.rs:171]: The problematic test_invalid_baseline_file has been removed and replaced with comprehensive error validation in test_baseline_error_validation_comprehensive.

File: tests/baseline.rs [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+#[tokio::test]
+async fn test_baseline_file_validation_failure() {
+    // Create a temporary invalid baseline file
+    let temp_path = "test_baseline_invalid.json";
+    create_test_baseline(Path::new(temp_path), false).expect("Failed to create test baseline");
+
+    // Test that invalid baseline fails during load test startup
+    let server = MockServer::start();
+    let mut configuration = common::build_configuration(&server, vec![]);
+
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply tests/baseline.rs:206]: The test_empty_baseline_file has been removed and replaced with test_baseline_empty_file which properly validates JSON parsing errors. The current test confirms that empty files are rejected during the parsing stage with appropriate error handling.

File: tests/baseline.rs [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+    fs::remove_file(temp_path).ok();
+}
+
+#[tokio::test]
+async fn test_baseline_file_not_found() {
+    // Test with non-existent file
+    let server = MockServer::start();
+    let mut configuration = common::build_configuration(&server, vec![]);
+
+    // Set non-existent baseline file directly on configuration
```
[the remark this answers is not in the document]
Comment (jeremyandrews) [reply tests/baseline.rs]: The test_malformed_json_baseline has been removed and the scenario is now covered by test_baseline_invalid_json. The current test properly validates that malformed JSON files are rejected during parsing with appropriate error messages.

**Comment by jeremyandrews** [comment]
> There is a significant amount of duplicated code in src/metrics/common.rs that should be refactored to improve maintainability.

This has been resolved in commit 0409014b which deduplicated the repeated deserialize implementations in src/metrics/common.rs.
